[WPE][Tools] cross-toolchain-helper: allow several image types per target and add support deploying those.

clopez opened the pull request:
#### e00e2b475934639bb1f92eaf85dde7b4906afd19
<pre>
[WPE][Tools] cross-toolchain-helper: allow several image types per target and add support deploying those.
<a href="https://bugs.webkit.org/show_bug.cgi?id=255861">https://bugs.webkit.org/show_bug.cgi?id=255861</a>

Reviewed by Philippe Normand.

This implements support for specifying more than one image type
per target and also supports automatic deploying of the images
from the script.

The cross-toolchain-helper now has the parameter `--deploy-image-with-script`
that allows to execute a script that will deploy the image generated.
Two examples scripts are included: one for deploying images of type
tarball to a local directory for remote booting (via NFS) and another
one to write the .wic image to a SDCard.

It also gets a new parameter `--check-if-image-is-updated` to verify
if the deployed or running image is updated to the last version.

This support for automatic deployment of images will be used on the
upcoming WPE performance bots.

* Tools/Scripts/cross-toolchain-helper:
* Tools/yocto/README.md:
* Tools/yocto/deploy_image_scripts/tarball-to-nfsdir.sh: Added.
* Tools/yocto/deploy_image_scripts/wic-to-sdcard.sh: Added.
* Tools/yocto/rpi/fix-nativesdk-image-defs.patch:
* Tools/yocto/rpi/manifest.xml:
* Tools/yocto/targets.conf:

Canonical link: <a href="https://commits.webkit.org/263861@main">https://commits.webkit.org/263861@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6394c84d0b1a0c49bc7419e1e25a0105be0ae23

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5369 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5506 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5695 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6907 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5405 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5369 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5733 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5485 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6207 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5468 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5538 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6935 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3017 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4815 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/11966 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4883 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4893 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6512 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5322 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4369 "8 flakes 1 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4786 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8883 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/695 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5147 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Tools/Scripts/cross-toolchain-helper:
```diff
@@ -605,27 +745,40 @@ def main(args):
+    if options.check_if_image_is_updated == 'running':
+        if options.target:
+            parser.error('Incompatible argument "--cross-target" with "--check-if-image-is-updated=running"')
+    else:
+        if not options.target:
```
`elif` ?

clopez replied:
There is another nested if blocks below (line 764), so using elif here would change the logic.

philn commented on Tools/Scripts/cross-toolchain-helper:
```diff
@@ -67,6 +67,11 @@ def configure_logging(selected_log_level='info'):
+def set_env_var(env_key, env_value, loginfo=False):
+    os.environ[env_key] = env_value
+    if loginfo:
+        _log.info('export {env_key}="{env_value}"'.format(env_key=env_key, env_value=env_value))
```
Side question, why using format strings in this stuff? Why not f-strings? Is this supposed to run with python2 which is EOL since Jan 1st 2020?

clopez replied:
Not good reason to do that, just a bad habit.
I would try to change the code to f-strings on an upcoming patch soon.

clopez committed:
[WPE][Tools] cross-toolchain-helper: allow several image types per target and add support deploying those.
https://bugs.webkit.org/show_bug.cgi?id=255861

Reviewed by Philippe Normand.

This implements support for specifying more than one image type
per target and also supports automatic deploying of the images
from the script.

The cross-toolchain-helper now has the parameter `--deploy-image-with-script`
that allows to execute a script that will deploy the image generated.
Two examples scripts are included: one for deploying images of type
tarball to a local directory for remote booting (via NFS) and another
one to write the .wic image to a SDCard.

It also gets a new parameter `--check-if-image-is-updated` to verify
if the deployed or running image is updated to the last version.

This support for automatic deployment of images will be used on the
upcoming WPE performance bots.

* Tools/Scripts/cross-toolchain-helper:
* Tools/yocto/README.md:
* Tools/yocto/deploy_image_scripts/tarball-to-nfsdir.sh: Added.
* Tools/yocto/deploy_image_scripts/wic-to-sdcard.sh: Added.
* Tools/yocto/rpi/fix-nativesdk-image-defs.patch:
* Tools/yocto/rpi/manifest.xml:
* Tools/yocto/targets.conf:

Canonical link: https://commits.webkit.org/263861@main
